### DISCUSSION TITLE
Time Zone Support

### DISCUSSION THREAD

**Discussion by hawkfish** [duckdb/duckdb#2647] [Ideas]
## TL;DR

We add the `TIMESTAMP WITH TIME ZONE` type to DuckDB and defer most of its manipulation to extensions such as `icu`.

## Introduction

Time zone support is a common request (e.g., #561) but it is complicated by the confusing nature of time and the SQL spec. I propose that we implement the functionality in two modular pieces:
* Add the SQL `TIME ZONE` types to the main DuckDB code base with minimal operations
* Extend the `icu` extension to implement the temporal operations.

This approach should make it possible to support time zones in a correct, extensible and mostly understandable manner.

## Definitions

Newtonian time is measured in _instants_. Instants are typically measured as a count of some duration from a fixed point in Newtonian space and time. In DuckDB, the fixed point is the Unix epoch of `1970-01-01 00:00:00` in UTC and the duration is microseconds (µs). Using this definition, a DuckDB `TIMESTAMP` column actually contains instants. Postgres and most other databases take this approach (although not MySQL, which uses a baroque system that allows for days and months of zero.)

For analysis and representation, time is _binned_ using a combination of a _calendar_ and a _time zone_. Binning is used for analysis by using functions such as `DATEPART` and `DATEDIFF`. It is used for representation when formatting dates for user consumption, such as ISO-8601 formats like `2021-11-21 07:24:21 -0800`. DuckDB uses the (proleptic) Gregorian calendar and the UTC time zone for these operations.

## Types

The SQL standard defines a number of time zone related types, which vary from confusing, through nearly useless, and ending up at ill-defined.

### TIMESTAMP WITH TIME ZONE

The most useful type defined by the standard is `TIMESTAMP WITH TIME ZONE`, which can be abbreviated `TIMEZONETZ`. This name is unfortunate because what it really seems to mean (at least in Postgres) is "instant binned using the current calendar and time zone settings of the connection". It does _not_ mean that a time zone or offset is stored with the data; just like a `TIMEZONE` column, a `TIMEZONETZ` column only contains instants.

What this does mean is that the two types contain the same values, but use different binning. This suggests that they can be cast to each other to alter the binning semantics. In order to avoid unexpected behaviour, the promotion cast here should be from `TIMEZONE` to `TIMEZONETZ`, and explicit downcasting can be specified in the other direction. This will allow ordinary `TIMEZONE`s to be used with `TIMEZONETZ` in instant calculations, but prevent existing functions from being inappropriately bound to `TIMEZONETZ`.

There are some operations on instants that can be implemented in the main DuckDB code. We can compare them for order and equality, find the difference between them in µs, and add µs to them. `INTERVAL` arithmetic is more problematic because `INTERVAL`s contain a mix of instant and calendar units. While it might seem draconian to not implement `TIMEZONETZ` + `INTERVAL` when the latter contains only instant units, it would make binding very confusing when a extension was added that implemented this operation for those types. Even worse, days are not even sized in the presence of daylight savings time. The proposal is therefore to not support this operator in the base code.

Since the base DuckDB code will not provide binning functions for `TIMEZONETZ`, they will need to be provided by a extension. In particular, the existing `icu` extension already has the data needed for the binning functions, and this will be described later.

Specifying `TIMEZONETZ` in SQL requires using a string representation. Such representations have an implied binning, which does not exist in the base code. This means that the only way to create `TIMEZONETZ` values is by explicitly casting a regular `TIMEZONE` value. While this requirement might seem excessive, it does force the user to understand what they are doing. Extensions that define string casting will then allow the user to specify a `TIMEZONETZ` value in a query.

For testing purposes, it might be helpful to provide a formatter if no cast is defined. If this causes too many problems, the test query can simply use downcasting to display the result in UTC.

### DATE WITH TIME ZONE

A less useful type defined by the standard is `DATE WITH TIME ZONE`, abbreviated `DATETZ`. A `DATETZ` is a `DATE` that is interpreted as a `TIMESTAMPTZ` at midnight of the date in question using the current binning.

The internal representation of `DATE` and `DATETZ` is a number of days since `1970-01-01`. Comparison is more problematic because changing the time zone can change the equality semantics. This suggests that implicit casting is not well-defined and should not be implemented. Explicit casting should be allowed, as it at least makes it clear to the user that they are doing something that is not well-defined.

Binning operations will need to be defined by a extension, and this will be discussed below. This includes string representations, and again, the only way to specify `DATETZ` values in the base code will be by casting `DATE`s.

### TIME WITH TIME ZONE

The `TIME WITH TIME ZONE` is not well-defined because binning a time requires a specific date as well as a time zone. We could decide to add this type for completeness, and it would have the same representation as `TIME`. The comparison semantic issues are the same as for `DATETZ` and can be overriden by explicit casts, but no implicit casting should be allowed. Extensions can add functionality if they have a semantically useful application, but we should not try to define them.

## Binning

DuckDB provides built in binning operations for instants, but the point of this feature is to extend binning functionality to arbitrary time zones. Using extensions, we can go beyond this and allow users to support arbitrary binning systems, including non-Gregorian calendars or even astronomical calendars that incorporate leap seconds.

### ICU

The most straightforward way to add time zones to DuckDB is to extend the existing `icu` extension, which already has the calendar and time zone data and functions. The extension would register new variants of the existing binning functions that bind to the `WITH TIME ZONE` types. These functions would reference settings defined by the extension, starting with `TimeZone`, and use the ICU functions for the computations. This includes existing functions such as `DATEPART` but could include new ones such as `ERA` or `ROLL`.

The settings used by `icu` need not be restricted to the time zone; the library also supports non-Gregorian calendars, so it could also allow a `Calendar` setting. In particular, this would add support for Japanese Imperial and Islamic Civil calendars. Formatting could also be overridden using the rich ICU format string descriptions instead of the simple `datestyle` setting provided by Postgres. The result would be to expose most of the ICU temporal functionality in DuckDB using an extension we already ship.

### Other Time Bases

A common warning in databases is that the calendar system is not appropriate for astronomical work, most notably because it does not consider leap-seconds and minutes with 61 seconds. With the extension approach, it would be possible to implement astronomical calendars by binning the instances using an appropriate library.

There may be other examples of time bases that could use this approach (e.g. satellite frame dragging computations) but the important point is that by having a well-defined internal representation of instants and a temporal type whose semantics can be defined by a user extension, we will have a powerful system for temporal analytics.

## Row Level Temporal Analytics

A final class of temporal queries are whose where the binning information is attached to particular rows:
* Queries where the local time of the actions are important for the analysis. An example might be whether a purchase was made in the early morning.
* Queries where the time at a physical plant is important for the analysis. An example here might be the local sun time for a solar power plant.

In the first  case, one might envision an enriched temporal data type that contained the time zone in addition to the instance. In the second case the location data would likely be joined to a second table of plant locations. The latter situation is more general, and it suggests that other columns in the data set could be used to drive the binning, and this could be accommodated by extending the binning functions to take the binning description as an extra parameter. Such extensions could be implemented simply by defining variants with more arguments, again via an extension.

**Comment by hannes** [comment]
I think the proposal is very sensible, indeed having the tz-types in the base system and the various interpretation and conversion functions in the ICU extension is the way to go. Two comments:

1) What should we do with the non-timezoned versions? Right now we just assume UTC and completely ignore the client time zone. Should this stay like this? Because now if you cast those to the annotated version surprising things (TM) happen because users will expect that non-annotated types will be somehow in their local time zone. 

2) For casting, we would need syntax to specify the target time zone, right? Should we copy Postgres and do something like `SET TIME ZONE 'Asia/Kathmandu';` and throw an error if it's not set? Perhaps it would be nice to be able to specify the time zone in a query itself, too, like `CAST(x AS TIMESTAMPTZ('Asia/Kathmandu')` or so.

**Reply by hawkfish** [reply]
I think the answer to 1 is "Yes, leave it alone" because that is what PG and the standard say. Adding an extension should not change the result of your bin calculation. All that happens with `TIMESTAMPTZ` with no extension is "function not defined".

As for 2, I was thinking we should check the ICU list (also for calendars, and date formats) and throw an exception if it is not valid.  I'm not sure if the current API has validation, but if not, we should really add that anyway.

**Reply by hawkfish** [reply]
Oh, the second part of 2 would be a new data type for the "row level analytics" because you are now attaching a binning system to a value. For that I was thinking one approach might be to use a `STRUCT` type with a `TIMESTAMP` and one or two other columns (time zone, calendar). Then we would need just extra binning functions.

**Comment by muschellij2** [comment]
In case anyone is coming to this from `R`, a workaround is that you need to send `LOAD icu` to your `duckdb` connection.

Below is a [`reprex`](https://reprex.tidyverse.org/) showing the issues with "TIMESTAMP WITH TIME ZONE" in R.

``` r
library(dbplyr)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:dbplyr':
#> 
#>     ident, sql
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(duckdb)
#> Loading required package: DBI
library(lubridate)
#> 
#> Attaching package: 'lubridate'
#> The following objects are masked from 'package:base':
#> 
#>     date, intersect, setdiff, union
library(arrow)
#> 
#> Attaching package: 'arrow'
#> The following object is masked from 'package:lubridate':
#> 
#>     duration
#> The following object is masked from 'package:utils':
#> 
#>     timestamp
library(DBI)
x = data.frame(
  time = Sys.time()
)

# I tried multiple timezone_out and tz_out_convert options, but always seemed to need `icu`
con <- DBI::dbConnect(duckdb::duckdb(), timezone_out = "")
tab <- to_duckdb(x, con, "data")
tab %>% 
  dplyr::mutate(time = lubridate::floor_date(time, "minutes")) %>% 
  dplyr::collect()
#> Error in `dplyr::collect()`:
#> ! Failed to collect lazy table.
#> Caused by error:
#> ! rapi_prepare: Failed to prepare query SELECT DATE_TRUNC('minutes', "time") AS "time"
#> FROM "data"
#> Error: Binder Error: No function matches the given name and argument types 'date_trunc(VARCHAR, TIMESTAMP WITH TIME ZONE)'. You might need to add explicit type casts.
#>  Candidate functions:
#>  date_trunc(VARCHAR, TIMESTAMP) -> TIMESTAMP
#>  date_trunc(VARCHAR, DATE) -> TIMESTAMP
#>  date_trunc(VARCHAR, INTERVAL) -> INTERVAL
#> 
#> LINE 1: SELECT DATE_TRUNC('minutes', "time") AS "time"
#>                ^
#> Backtrace:
#>     ▆
#>  1. ├─... %>% dplyr::collect()
#>  2. ├─dplyr::collect(.)
#>  3. └─dbplyr:::collect.tbl_sql(.)
#>  4.   └─base::tryCatch(...)
#>  5.     └─base (local) tryCatchList(expr, classes, parentenv, handlers)
#>  6.       └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
#>  7.         └─value[[3L]](cond)
#>  8.           └─cli::cli_abort("Failed to collect lazy table.", parent = cnd)
#>  9.             └─rlang::abort(...)
# need for TIMEZONE WITH
# run this once to install icu on duckdb locally R
# dbSendQuery(con, sql("INSTALL icu; LOAD icu;"))
dbSendQuery(con, sql("LOAD icu;"))
#> <duckdb_result 82c70 connection=6fd90 statement='LOAD icu;'>
tab %>% 
  dplyr::mutate(time = lubridate::floor_date(time, "minutes")) %>% 
  dplyr::collect()
#> # A tibble: 1 × 1
#>   time               
#>   <dttm>             
#> 1 2023-08-29 16:59:00
```

<sup>Created on 2023-08-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.1 (2022-06-23)
#>  os       macOS Big Sur ... 10.16
#>  system   x86_64, darwin17.0
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/New_York
#>  date     2023-08-29
#>  pandoc   3.1.5 @ /usr/local/bin/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version  date (UTC) lib source
#>  arrow       * 12.0.1.1 2023-07-18 [1] CRAN (R 4.2.0)
#>  assertthat    0.2.1    2019-03-21 [1] CRAN (R 4.2.0)
#>  bit           4.0.5    2022-11-15 [1] CRAN (R 4.2.0)
#>  bit64         4.0.5    2020-08-30 [1] CRAN (R 4.2.0)
#>  blob          1.2.4    2023-03-17 [1] CRAN (R 4.2.0)
#>  cli           3.6.1    2023-03-23 [1] CRAN (R 4.2.0)
#>  DBI         * 1.1.3    2022-06-18 [1] CRAN (R 4.2.0)
#>  dbplyr      * 2.3.3    2023-07-07 [1] CRAN (R 4.2.0)
#>  digest        0.6.33   2023-07-07 [1] CRAN (R 4.2.0)
#>  dplyr       * 1.1.2    2023-04-20 [1] CRAN (R 4.2.0)
#>  duckdb      * 0.8.1-1  2023-07-17 [1] CRAN (R 4.2.0)
#>  evaluate      0.21     2023-05-05 [1] CRAN (R 4.2.0)
#>  fansi         1.0.4    2023-01-22 [1] CRAN (R 4.2.0)
#>  fastmap       1.1.1    2023-02-24 [1] CRAN (R 4.2.0)
#>  fs            1.6.3    2023-07-20 [1] CRAN (R 4.2.0)
#>  generics      0.1.3    2022-07-05 [1] CRAN (R 4.2.0)
#>  glue          1.6.2    2022-02-24 [1] CRAN (R 4.2.0)
#>  htmltools     0.5.5    2023-03-23 [1] CRAN (R 4.2.0)
#>  knitr         1.43     2023-05-25 [1] CRAN (R 4.2.0)
#>  lifecycle     1.0.3    2022-10-07 [1] CRAN (R 4.2.0)
#>  lubridate   * 1.9.2    2023-02-10 [1] CRAN (R 4.2.0)
#>  magrittr      2.0.3    2022-03-30 [1] CRAN (R 4.2.0)
#>  pillar        1.9.0    2023-03-22 [1] CRAN (R 4.2.0)
#>  pkgconfig     2.0.3    2019-09-22 [1] CRAN (R 4.2.0)
#>  purrr         1.0.1    2023-01-10 [1] CRAN (R 4.2.0)
#>  R.cache       0.16.0   2022-07-21 [1] CRAN (R 4.2.0)
#>  R.methodsS3   1.8.2    2022-06-13 [1] CRAN (R 4.2.0)
#>  R.oo          1.25.0   2022-06-12 [1] CRAN (R 4.2.0)
#>  R.utils       2.12.2   2022-11-11 [1] CRAN (R 4.2.0)
#>  R6            2.5.1    2021-08-19 [1] CRAN (R 4.2.0)
#>  reprex        2.0.2    2022-08-17 [1] CRAN (R 4.2.0)
#>  rlang         1.1.1    2023-04-28 [1] CRAN (R 4.2.0)
#>  rmarkdown     2.23     2023-07-01 [1] CRAN (R 4.2.0)
#>  rstudioapi    0.15.0   2023-07-07 [1] CRAN (R 4.2.0)
#>  sessioninfo   1.2.2    2021-12-06 [1] CRAN (R 4.2.0)
#>  styler        1.10.1   2023-06-05 [1] CRAN (R 4.2.0)
#>  tibble        3.2.1    2023-03-20 [1] CRAN (R 4.2.0)
#>  tidyselect    1.2.0    2022-10-10 [1] CRAN (R 4.2.0)
#>  timechange    0.2.0    2023-01-11 [1] CRAN (R 4.2.0)
#>  utf8          1.2.3    2023-01-31 [1] CRAN (R 4.2.0)
#>  vctrs         0.6.3    2023-06-14 [1] CRAN (R 4.2.0)
#>  withr         2.5.0    2022-03-03 [1] CRAN (R 4.2.0)
#>  xfun          0.39     2023-04-20 [1] CRAN (R 4.2.0)
#>  yaml          2.3.7    2023-01-23 [1] CRAN (R 4.2.0)
#> 
#>  [1] /Library/Frameworks/R.framework/Versions/4.2/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

**Comment by hawkfish** [comment]
Closing because this [has been implemented](https://duckdb.org/2022/01/06/time-zones.html) for some time now.

**Comment by muschellij2** [comment]
@hawkfish - I agree, but I was surprised when the command for `R` failed out of the box for the time, without previous timezone setting or using of `icu` and wanted to give a solution for R users who may encounter same issue.
